### PR TITLE
Fix #367

### LINKED ISSUE #367: Peer table (node window) does not remain sorted on update
image

This issue was introduced somewhere between 672870ab7ba51b1fc40f263a1c71ebcf165b1229 and c8571486364d6e9ca8c86bd1c81e230ca64f8904

### PR DESCRIPTION
**Pull request by rebroad**
Attempting to fix issue #367

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Please explain the change and remove the unscoped GitHub attribute. The unscoped link will mess with all monotree repos, not just this one.

**Comment by rebroad** [comment]
@MarcoFalke what is an unscoped github attribute please?

**Comment by MarcoFalke** [comment]
`#367` is unscoped. At the very least it should say `bitcoin-core/gui#374`.

**Comment by rebroad** [comment]
> `#367` is unscoped. At the very least it should say `bitcoin-core/gui#374`.

A pull/issue always defaults to the same project that it's in - it's only necessary to add the project if it's a different project to that which the pull/issue is in.

**Comment by rebroad** [comment]
How do I change this to a draft pull request? I think there may be a chance of Segmentation fault in this code.

**Comment by MarcoFalke** [comment]
> A pull/issue always defaults to the same project that it's in - it's only necessary to add the project if it's a different project to that which the pull/issue is in.

This is a monotree project. Meaning that different project will have the same tree and share the (merge) commits. So without a scope it won't be trivially possible to recover the metadata.

**Comment by fanquake** [comment]
> A pull/issue always defaults to the same project that it's in - it's only necessary to add the project if it's a different project to that which the pull/issue is in.

Rather than arguing the point, how about just updating the link. As mentioned, having an unscoped link will mess with our merge process / resultant commit messages.

> How do I change this to a draft pull request?

Click the button that says "Convert to draft"

draft

You do realise you could read the GitHub documentation, or use any search engine to also find the answer to that question? Rather than posting another comment here?

I'd suggest you spend some time reading this projects developer docs and contributing guide. Opening a PR with no title, no explanation of the change, a useless commit message, and no commit body is not the standard that is expected here. I'm also almost certain you are aware of that, given you've been around the project for many years now.

**Review by hebasto — COMMENTED** [review]
Tested 0d41edb8b3ce6923821985a633b49c3be002a05b on Linux Mint 20.1 (Qt 5.12.8) -- works as intended.

The change is correct (UPDATE: see https://github.com/bitcoin-core/gui/pull/374#issuecomment-869662723), and Qt [docs](https://doc.qt.io/qt-5/model-view-programming.html#resizable-models) clearly states:
> For more complex changes to the model's structure, perhaps involving internal reorganization, _sorting of data_ or any other structural change, it is necessary to perform the following sequence:
>
>  - Emit the layoutAboutToBeChanged() signal
>  - Update internal data which represents the structure of the model.
>  - Update persistent indexes using changePersistentIndexList()
>  - Emit the layoutChanged() signal.

UPDATE 2: see https://github.com/bitcoin-core/gui/pull/374#issuecomment-869744938

Also this signal seems redundant now:
```diff
--- a/src/qt/peertablemodel.h
+++ b/src/qt/peertablemodel.h
@@ -73,9 +73,6 @@ public:
 public Q_SLOTS:
     void refresh();
 
-Q_SIGNALS:
-    void changed();
-
 private:
     //! Internal peer data structure.
     QList<CNodeCombinedStats> m_peers_data{};
```

Hope that aforementioned comments will be addressed as well.

**Comment by hebasto** [comment]
OTOH, this change re-introduces #191 bug.

**Comment by rebroad** [comment]
> OTOH, this change re-introduces #191 bug.

Out of the two bugs #191 or #367 - I'd rather have former than the latter.

**Comment by MarcoFalke** [comment]
> I'm trying to understand the issue. I can't fix something when I'm not aware of the problem.

There are two issues with the commit message:

* The hashtag is missing the repo namespace, so when the commit gets pushed to another repo it can't be resolved properly.
* The commit message is lacking any sort of explanation.

You can read our contributing guide to learn on how to create patches:  https://github.com/bitcoin-core/gui/blob/master/CONTRIBUTING.md#committing-patches . It also links to a guide on how to write commit messages.

**Comment by hebasto** [comment]
I still don't understand why [this](https://doc.qt.io/qt-5/qsortfilterproxymodel.html#details) does not work:
> By default, the model dynamically re-sorts and re-filters data whenever the original model changes.

And we have `dynamicSortFilter == true` for sure.

**Comment by rebroad** [comment]
> * The hashtag is missing the repo namespace, so when the commit gets pushed to another repo it can't be resolved properly.

Ah, I understand now - you're implying that it will also get pushed to https://github.com/bitcoin/bitcoin, yes?

> * The commit message is lacking any sort of explanation.

"Fixes #367" is an explanation. so I would disagree with your statement, but if it gets pushed to another repository then I can understand how it could point people to the wrong information. I'm not really understanding why we have two repositories with duplicate code though. Is there an explanation for this somewhere?

> You can read our contributing guide to learn on how to create patches: https://github.com/bitcoin-core/gui/blob/master/CONTRIBUTING.md#committing-patches . It also links to a guide on how to write commit messages.

Thanks. I shall read it now.

**Comment by hebasto** [comment]
My previous [comment](https://github.com/bitcoin-core/gui/pull/374#pullrequestreview-693918856) about using of `layoutAboutToBeChanged` and `layoutChanged` signals is not correct because it should be applied to cases when no proxy models are used.

I believe that correct fix of #367 is suggested in #375.

**Comment by hebasto** [comment]
Does 71ab42d00e12c3bbdf3ca588da6408ec209a10d8 actually revert ecbd91153875c8cdd5b92b840afc116f65e457fb?

File: src/qt/peertablemodel.cpp [hunk 1]
```diff
@@ -154,30 +154,7 @@ void PeerTableModel::refresh()
         new_peers_data.append(stats);
     }
 
-    // Handle peer addition or removal as suggested in Qt Docs. See:
```
Comment (jarolrod) [inline src/qt/peertablemodel.cpp:157]: removing this reintroduces #160 as described here: https://github.com/bitcoin-core/gui/issues/160#issuecomment-855515389

**Review by jarolrod — COMMENTED** [review]
NACK

This is not the way to fix the mentioned issue. This also reintroduces two bugs that we had fixed (#160 and #191)

**Comment by hebasto** [comment]
Closing due to the #375 is merged.

**Pull request closed without merging**